velox/functions/lib/Re2Functions.cpp

```diff
@@ -966,56 +1018,141 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
   };
 }
 
-std::string unescape(
-    std::string_view pattern,
+// Kind of sub-pattern.
+enum class SubPatternKind {
+  /// e.g. '___'.
+  kSingleCharWildcard,
+  // e.g. '%%'.
+  kAnyCharsWildcard,
+  // e.g. 'abc'.
+  kLiteralString
+};
+
+// Sub-pattern's stats in a top-level pattern.
+struct SubPatternStats {
+  SubPatternKind kind;
+  // Count of this patterns in total.
+  size_t count = 0;
+  // First index of this pattern kind.
+  int32_t firstIndex = -1;
+  // Last index of this pattern kind.
+  int32_t lastIndex = -1;
+
+  void update(size_t index) {
+    count++;
+    if (firstIndex == -1) {
+      firstIndex = index;
+      lastIndex = index;
```

Comment:
this code is the same in both if and else branches; consider moving it outside of the if-else block